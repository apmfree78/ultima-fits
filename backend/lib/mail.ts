import { createTransport, getTestMessageUrl } from 'nodemailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: process.env.MAIL_PORT,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeANiceEmail(text: string): string {
  return `
   <div style="
    border: 1px solid black;
    padding:20px;
    font-family: sans-sarif;
    line-height: 2;
    font-size: 20px;
    ">
    <h2>Hello There!</h2>
    <p>${text}</p>
    <p>Amit Mehta</p>

   </div>
  `;
}

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  // email user a token
  const info = await transport.sendMail({
    to,
    from: 'amit@profitswami.com',
    subject: 'Your Password Reset Token',
    html: makeANiceEmail(`Your Password Reset Token is here!
    
    <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click Here to Reset</a>`),
  });
  if (process.env.MAIL_USER.includes('ethereal.email')) {
    console.log(`Message Sent! Preview it at ${getTestMessageUrl(info)}`);
  }
}
