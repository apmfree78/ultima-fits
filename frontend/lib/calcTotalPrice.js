export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    if (!cartItem.product) return tally; // products can be delete but can still be in the cart
    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
}
