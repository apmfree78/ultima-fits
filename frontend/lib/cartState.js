/* eslint-disable react/prop-types */
import { createContext, useContext, useState } from 'react';

const LocalStateContent = createContext();
const LocalStateProvider = LocalStateContent.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider!
  // We will store data (state) and functionality (updaters) in here
  // and anyone can access it via the consumer!

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, openCart, closeCart, toggleCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for accessing the cart local state
function useCart() {
  // we use a consumer here to access the local state
  const all = useContext(LocalStateContent);
  return all;
}
export { CartStateProvider, useCart };
