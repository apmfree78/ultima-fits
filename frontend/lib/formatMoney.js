export default function FormatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
  };
  const formatter = Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}
