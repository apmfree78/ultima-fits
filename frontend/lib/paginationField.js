/* eslint-disable no-plusplus */
import { PAGINATION_QUERY } from '../components/pagination';

export default function paginationField() {
  return {
    keyArgs: false,
    read(existing = [], { args, cache }) {
      console.log({ existing, args, cache });
      const { skip, first } = args;

      // read the number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check if you have existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      // if there are items AND there aren't enough items
      // AND we are on the last page THEN JUST SENT IT
      if (items.length && items.length !== first && page === pages) {
        return items;
      }

      if (items.length !== first) {
        // we don't have any items, we must go to network to fetch them
        return false;
      }

      // if there are items, then return them from cache
      if (items.length) {
        console.log(
          `there are ${items.length} items in the cache! Gonna sent em over to apollo`
        );
        return items;
      }

      return false; // fall back to network

      // asks read functions for items
      // 1st thing we can do is return the items
      // because they are already in the cache
      // the other thing we can do is return false from here
      // which will result in a network request
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      // this runs when the apollo client
      // comes back from the network with products
      console.log(`merging items from the network ${incoming.length}`);
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
