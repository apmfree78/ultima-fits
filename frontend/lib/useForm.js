import { useEffect, useState } from 'react';

export default function useForm(initial = {}) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    // this functions runs when things we are watching change
    setInputs(initial);
  }, [initialValues]);

  function handleChange(e) {
    let { value, name, type } = e.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = e.target.files;
    }
    setInputs({
      // copy existing state
      ...inputs,
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }
  // return the things we want to surface from custom hook

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
