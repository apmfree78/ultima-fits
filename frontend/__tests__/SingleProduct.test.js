import { MockedProvider } from '@apollo/react-testing';
import { render, screen } from '@testing-library/react';
import Product from '../components/Product';
import SingleProduct, { SINGLE_ITEM_QUERY } from '../components/SingleProduct';
import { fakeItem } from '../lib/testUtils';

const product = fakeItem();

const mocks = [
  {
    // When some one reqests this query and variable combo
    request: {
      query: SINGLE_ITEM_QUERY,
      variables: {
        id: '123',
      },
    },
    // return this data
    result: {
      data: {
        Product: product,
      },
    },
  },
];

describe('<SingleProduct/>', () => {
  it('renders with proper data', async () => {
    // we need to make some fake data
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    // wait for the test id to show up
    await screen.findByTestId('singleProduct');
    debug();
    expect(container).toMatchSnapshot();
  });

  it('Errors out when an item is not found', async () => {
    const ErrorMock = [
      {
        request: {
          query: SINGLE_ITEM_QUERY,
          variables: {
            id: '123',
          },
        },
        result: {
          errors: [{ message: 'Item not found!' }],
        },
      },
    ];
    const { container, debug } = render(
      <MockedProvider mocks={ErrorMock}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    await screen.findByTestId('graphql-error');
    debug();
    expect(container).toHaveTextContent('Shoot!');
    expect(container).toHaveTextContent('Item not found!');
  });
});
