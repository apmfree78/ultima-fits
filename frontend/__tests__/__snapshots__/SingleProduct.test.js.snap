// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct/> renders with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-1kapz8w-0 dIOaRa"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
